snapcraft/storeapi/_ubuntu_sso_client.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+
+import logging
+import json
+import os
+from typing import Optional, TextIO
+from urllib.parse import urljoin, urlparse
+
+import pymacaroons
+import requests
+from requests.adapters import HTTPAdapter
+from requests.exceptions import ConnectionError, RetryError
+from requests.packages.urllib3.util.retry import Retry
+from simplejson.scanner import JSONDecodeError
+
+from snapcraft import config
+from . import _agent, constants, errors
+
+
+logger = logging.getLogger(__name__)
+
+# Set urllib3's logger to only emit errors, not warnings. Otherwise even
+# retries are printed, and they're nasty.
+logging.getLogger(requests.packages.urllib3.__package__).setLevel(logging.ERROR)
+
+
+class Client:
+    """Generic Client to talk to the *Store."""
+
+    def __init__(self, *, user_agent: str = _agent.get_user_agent()) -> None:
+        self.session = requests.Session()
+        self._user_agent = user_agent
+
+        # Setup max retries for all store URLs and the CDN
+        retries = Retry(
+            total=int(os.environ.get("STORE_RETRIES", 5)),
+            backoff_factor=int(os.environ.get("STORE_BACKOFF", 2)),
+            status_forcelist=[104, 500, 502, 503, 504],
+        )
+        self.session.mount("http://", HTTPAdapter(max_retries=retries))
+        self.session.mount("https://", HTTPAdapter(max_retries=retries))
+
+    def request(
+        self, method, url, params=None, headers=None, **kwargs
+    ) -> requests.Response:
+        """Send a request to url relative to the root url.
+
+        :param str method: Method used for the request.
+        :param str url: Appended with the root url first.
```

Comment:
This parameter description sounds a bit confusing, we should append (prepend?) the root url or it will be done automatically for us?

Reply:
This is probably a copy paste error, I will update